src/test/java/org/jenkinsci/plugins/NodePreProvisionTest.java

```diff
@@ -0,0 +1,109 @@
+import java.util.logging.Handler;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
+
+import org.jenkinsci.plugins.vSphereCloud;
+import org.jenkinsci.plugins.vSphereCloudSlaveTemplate;
+import org.jenkinsci.plugins.vSphereCloud.VSpherePlannedNode;
+import org.jenkinsci.plugins.vsphere.VSphereConnectionConfig;
+import org.jenkinsci.plugins.vsphere.tools.CloudProvisioningAlgorithm;
+import org.jenkinsci.plugins.vsphere.tools.CloudProvisioningRecord;
+import org.jenkinsci.plugins.vsphere.tools.CloudProvisioningState;
+
+import hudson.model.Label;
+import hudson.slaves.JNLPLauncher;
+import hudson.slaves.RetentionStrategy;
+import hudson.slaves.NodeProvisioner.PlannedNode;
+
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class NodePreProvisionTest {
+    /** Used when faking up test data */
+    private static List<vSphereCloudSlaveTemplate> stubVSphereCloudTemplates;
+    private static VSphereConnectionConfig vsConnectionConfig;
+    private static vSphereCloud stubVSphereCloud;
+    private static CloudProvisioningState stubVSphereTemplateState;
+    private Logger testLogger;
+    private List<LogRecord> loggedMessages;
+
+    @BeforeClass
+    public static void setupClass() {
+        stubVSphereCloudTemplates = new ArrayList<vSphereCloudSlaveTemplate>();
+        vsConnectionConfig = new VSphereConnectionConfig("vsHost", false, "credentialsId");
+        stubVSphereCloud = new vSphereCloud(vsConnectionConfig, "vsDescription", 100, 100, stubVSphereCloudTemplates);
+        stubVSphereTemplateState = new CloudProvisioningState(stubVSphereCloud);
+    }
+
+    @Before
+    public void setup() {
+        stubVSphereCloudTemplates.clear();
+        loggedMessages = new ArrayList<LogRecord>();
+        // Get vSphereCloud logger
+        Logger logger = Logger.getLogger("vsphere-cloud");
+        logger.setLevel(Level.ALL);
+        // final Handler[] handlers = logger.getHandlers();
+        // for (final Handler handler : handlers) {
+        //     logger.removeHandler(handler);
+        // }
+        final Handler testHandler = new Handler() {
+            @Override
+            public void publish(LogRecord record) {
+                loggedMessages.add(record);
+            }
+
+            @Override
+            public void flush() {
+            }
+
+            @Override
+            public void close() {
+            }
+        };
+        logger.addHandler(testHandler);
+        testLogger = logger;
+    }
+
+    @Test 
+    public void shouldPreProvisionNodesWhenNotEnough() {
+        vSphereCloudSlaveTemplate template = createTemplate(10, 2);
+        provisionNode(template);
+        assertThat(stubVSphereTemplateState.countNodes(), equalTo(1));
+
+        // Here it says that there still should be provisioned 2 nodes, despite the fact there is 1 active already
+        stubVSphereCloud.preProvisionNodes(template);
```

Comment:
The logs: 
```
Jun 01, 2021 5:54:23 PM org.jenkinsci.plugins.vSphereCloud InternalLog
INFO: STARTING VSPHERE CLOUD

Jun 01, 2021 5:54:23 PM org.jenkinsci.plugins.vsphere.tools.CloudProvisioningState logStateChange
WARNING: Marking templateCapacity10instanceMin21 as active : wasPreviouslyPlanned!=true

Jun 01, 2021 5:54:24 PM org.jenkinsci.plugins.vSphereCloud preProvisionNodes
INFO: preProvisionNodesForTemplate(null): should pre-provision 2 nodes

Jun 01, 2021 5:54:24 PM org.jenkinsci.plugins.vSphereCloud calculateMaxAdditionalSlavesPermitted
INFO: There are 0 VMs in this cloud. The instance cap for the cloud is 100, so we have room for more

Jun 01, 2021 5:54:24 PM org.jenkinsci.plugins.vSphereCloud calculateMaxAdditionalSlavesPermitted
INFO: There are 1 VMs in this cloud. The instance cap for the cloud is 100, so we have room for more
```

Reply:
Note,  that when it preProvision is invoked it actually sees that "There are 1 VMs in this cloud", which was "provisioned" by the method just a moment ago.
So, I would love to achieve the same state without calling preProvision(); I've tried marking the associated record as `provisionedSlaveNowActive` or `provisioningStarted`, but that doesn't do the trick, neither calling `VSpherePlannedNode.createInstance(stubVSphereTemplateState, nodeName, provisionable);`